### PR TITLE
Feat(connector): [Fiuu] Add DuitNow/FPX PaymentMethod

### PR DESCRIPTION
**Pull request by awasthi21**
## Type of Change

- [ ] Bugfix
- [x] New feature
- [ ] Enhancement
- [x] Refactoring
- [ ] Dependency updates
- [ ] Documentation
- [ ] CI/CD

## Description

Added DuitNowQR 
Added FPX Payment Method

### Additional Changes

- [ ] This PR modifies the API contract
- [ ] This PR modifies the database schema
- [x] This PR modifies application configuration/environment variables

## How did you test it?
**Request**
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key:' \
--data '{
    "amount": 1222,
    "currency": "MYR",
    "confirm": true,
    "capture_method": "manual",
    "return_url": "https://google.com",
    "authentication_type": "no_three_ds",
    "payment_method": "real_time_payment",
    "payment_method_type": "duit_now",
    "payment_method_data": {
        "real_time_payment": {
            "duit_now":{
            }
        }
    },
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS"
}'
```
Response
```
{
    "payment_id": "pay_9l079slQTfxcnWXcvVWv",
    "merchant_id": "postman_merchant_GHAction_ff03d9c9-f319-4cf4-b657-f45aefd4cf5b",
    "status": "requires_customer_action",
    "amount": 1222,
    "net_amount": 1222,
    "amount_capturable": 1222,
    "amount_received": null,
    "connector": "fiuu",
    "client_secret": "pay_9l079slQTfxcnWXcvVWv_secret_eLSU6KF41kqq8PFWzEDZ",
    "created": "2024-09-09T11:59:34.213Z",
    "currency": "MYR",
    "customer_id": null,
    "customer": null,
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "manual",
    "payment_method": "real_time_payment",
    "payment_method_data": {
        "real_time_payment": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": null,
    "name": null,
    "phone": null,
    "return_url": "https://google.com/",
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": "joseph",
    "statement_descriptor_suffix": "JS",
    "next_action": {
        "type": "qr_code_information",
        "image_data_url": "data:image/png;base64",
        "display_to_timestamp": null,
        "qr_code_url": null
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "duit_now",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": null,
    "connector_transaction_id": null,
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_glzU2Ijq0fPKjPpGv4VV",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_hqcC5QIpdCOQMXIKCzum",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-09-09T12:14:34.213Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-09-09T11:59:35.101Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null
}
```
image

**FPX **
Request
```
curl --location 'http://localhost:8080/payments' \
--header 'Content-Type: application/json' \
--header 'Accept: application/json' \
--header 'api-key: dev_P5gdeka6FUxRQ04KOCTziGwl3d7RmGAhi6SIYqSyf8AOcvoAG9yVbdnH5I14b8H7' \
--data '{
  "amount": 1000,
  "currency": "MYR",
  "amount_to_capture":1000,
  "confirm": true,
  "capture_method": "automatic",
  "payment_method": "bank_redirect",
    "payment_method_type": "online_banking_fpx",
    "payment_method_data": {
      "bank_redirect": {
        "online_banking_fpx": {
          "issuer": "bank_islam"
        }
      }
    }
}'
```
Response
```
{
    "payment_id": "pay_NHJitFfMFucT4YlvLPql",
    "merchant_id": "postman_merchant_GHAction_ef7eaa6b-1a6d-47e0-8922-48ef5fd2058d",
    "status": "requires_customer_action",
    "amount": 1000,
    "net_amount": 1000,
    "amount_capturable": 1000,
    "amount_received": null,
    "connector": "fiuu",
    "client_secret": "pay_NHJitFfMFucT4YlvLPql_secret_S26ts7Aptec4Phq9ng4y",
    "created": "2024-09-10T12:49:44.741Z",
    "currency": "MYR",
    "customer_id": null,
    "customer": null,
    "description": null,
    "refunds": null,
    "disputes": null,
    "mandate_id": null,
    "mandate_data": null,
    "setup_future_usage": null,
    "off_session": null,
    "capture_on": null,
    "capture_method": "automatic",
    "payment_method": "bank_redirect",
    "payment_method_data": {
        "bank_redirect": {},
        "billing": null
    },
    "payment_token": null,
    "shipping": null,
    "billing": null,
    "order_details": null,
    "email": null,
    "name": null,
    "phone": null,
    "return_url": null,
    "authentication_type": "no_three_ds",
    "statement_descriptor_name": null,
    "statement_descriptor_suffix": null,
    "next_action": {
        "type": "redirect_to_url",
        "redirect_to_url": "http://localhost:8080/payments/redirect/pay_NHJitFfMFucT4YlvLPql/postman_merchant_GHAction_ef7eaa6b-1a6d-47e0-8922-48ef5fd2058d/pay_NHJitFfMFucT4YlvLPql_1"
    },
    "cancellation_reason": null,
    "error_code": null,
    "error_message": null,
    "unified_code": null,
    "unified_message": null,
    "payment_experience": null,
    "payment_method_type": "online_banking_fpx",
    "connector_label": null,
    "business_country": null,
    "business_label": "default",
    "business_sub_label": null,
    "allowed_payment_method_types": null,
    "ephemeral_key": null,
    "manual_retry_allowed": null,
    "connector_transaction_id": "30873605",
    "frm_message": null,
    "metadata": null,
    "connector_metadata": null,
    "feature_metadata": null,
    "reference_id": null,
    "payment_link": null,
    "profile_id": "pro_sBgvmr7vRtnw4EVuvg72",
    "surcharge_details": null,
    "attempt_count": 1,
    "merchant_decision": null,
    "merchant_connector_id": "mca_K6qyBEGB0tOHdbppPILi",
    "incremental_authorization_allowed": null,
    "authorization_count": null,
    "incremental_authorizations": null,
    "external_authentication_details": null,
    "external_3ds_authentication_attempted": false,
    "expires_on": "2024-09-10T13:04:44.741Z",
    "fingerprint": null,
    "browser_info": null,
    "payment_method_id": null,
    "payment_method_status": null,
    "updated": "2024-09-10T12:49:45.358Z",
    "charges": null,
    "frm_metadata": null,
    "merchant_order_reference_id": null
}
```
## Checklist

- [x] I formatted the code `cargo +nightly fmt --all`
- [x] I addressed lints thrown by `cargo clippy`
- [x] I reviewed the submitted code
- [ ] I added unit tests for my changes where possible

### CODE REVIEW

File: config/config.example.toml [hunk 1]
```diff
@@ -209,6 +209,7 @@ fiserv.base_url = "https://cert.api.fiservapps.com/"
 fiservemea.base_url = "https://prod.emea.api.fiservapps.com/sandbox"
 fiuu.base_url = "https://sandbox.merchant.razer.com/"
 fiuu.secondary_base_url="https://sandbox.merchant.razer.com/"
+fiuu.third_base_url="https://api.merchant.razer.com/"
```
Comment (deepanshu-iiitu) [inline config/config.example.toml:212]: Add this in `loadtest/config/development.toml`

File: crates/hyperswitch_connectors/src/utils.rs [hunk 2]
```diff
@@ -1529,3 +1531,37 @@ pub trait ForeignTryFrom<F>: Sized {
 
     fn foreign_try_from(from: F) -> Result<Self, Self::Error>;
 }
+
+#[derive(Debug)]
+pub struct QrImage {
+    pub data: String,
+}
```
Comment (deepanshu-iiitu) [inline crates/hyperswitch_connectors/src/utils.rs:1538]: Remove this struct from crate router and reexport from here

File: crates/hyperswitch_connectors/src/utils.rs [hunk 3]
```diff
@@ -1529,3 +1531,37 @@ pub trait ForeignTryFrom<F>: Sized {
...
+pub(crate) const QR_IMAGE_DATA_SOURCE_STRING: &str = "data:image/png;base64";
+pub(crate) const BASE64_ENGINE: base64::engine::GeneralPurpose =
```
Comment (deepanshu-iiitu) [inline crates/hyperswitch_connectors/src/utils.rs:1540]: use `BASE64_ENGINE` which is already present at `crates/common_utils/src/consts.rs` and remove this declaration

File: crates/hyperswitch_connectors/src/connectors/fiuu/transformers.rs [hunk 4]
```diff
@@ -877,3 +965,31 @@ impl From<RefundStatus> for enums::RefundStatus {
         }
     }
 }
+#[derive(Debug, Serialize, Deserialize)]
+pub struct DuitNowQrCodeResponse {
+    status: bool,
+    qrcode_data: Secret<String>,
+}
```
Comment (deepanshu-iiitu) [inline crates/hyperswitch_connectors/src/connectors/fiuu/transformers.rs:972]: Move this struct to line 235 of this file

**Review by deepanshu-iiitu — CHANGES_REQUESTED** [review]

**Review by deepanshu-iiitu — DISMISSED** [review]

File: crates/hyperswitch_connectors/Cargo.toml [hunk 5]
```diff
@@ -19,7 +19,9 @@ bytes = "1.6.0"
 error-stack = "0.4.1"
 hex = "0.4.3"
 http = "0.2.12"
+image = { version = "0.25.1", default-features = false, features = ["png"] }
 once_cell = "1.19.0"
+qrcode = "0.14.0"
```
Comment (SanchithHegde) [inline crates/hyperswitch_connectors/Cargo.toml:24]: Remove these 2 dependencies from the `router` crate if it no longer needs them.

**Review by SanchithHegde — APPROVED** [review]

**Review by deepanshu-iiitu — APPROVED** [review]

**Pull request merged**